### PR TITLE
Decouple persistent state and undoable state

### PR DESCRIPTION
**Pull request by felixcodes**
#483 

- Separate persistent state from undoable state

### CODE REVIEW

File: src/components/data-pane/data-selector.tsx [hunk 1]
```diff
@@ -188,9 +188,10 @@ const DataSelectorRenderer = CSSModules(DataSelectorBase, styles);
 
 export const DataSelector = connect(
   (state: State) => {
+    const presentUndoableState = state.undoable.present;
     return {
-      data: state.present.dataset,
-      name: state.present.dataset.name,
+      data: presentUndoableState.dataset,
```
Comment (kanitw) [inline src/components/data-pane/data-selector.tsx]: We should use a selector for dataset for consistency and for reduce the amount of updates needed if we reshape the state in the future.

File: src/components/data-pane/data-selector.tsx [hunk 2]
```diff
@@ -188,9 +188,10 @@ const DataSelectorRenderer = CSSModules(DataSelectorBase, styles);
...
+      name: presentUndoableState.dataset.name,
```
Comment (kanitw) [inline src/components/data-pane/data-selector.tsx]: It's weird that we need name since it's already a part of data.  File an issue and submit a separate PR?

File: src/components/data-pane/index.tsx [hunk 3]
```diff
@@ -47,9 +47,10 @@ export class DataPanelBase extends React.PureComponent<DataPanelProps, {}> {
 
 export const DataPane = connect(
   (state: State) => {
+    const presentUndoableState = state.undoable.present;
     return {
-      data: state.present.dataset,
-      config: state.present.config
+      data: presentUndoableState.dataset,
```
Comment (kanitw) [inline src/components/data-pane/index.tsx]: We should use a selector for dataset, config for consistency and for reduce the amount of updates needed if we reshape the state in the future.

File: src/components/encoding-pane/index.tsx [hunk 4]
```diff
@@ -162,12 +162,13 @@ class EncodingPanelBase extends React.PureComponent<EncodingPanelProps, {}> {
 
 export const EncodingPane = connect(
   (state: State) => {
+    const presentUndoableState = state.undoable.present;
     return {
-      spec: state.present.shelf.spec,
-      filters: state.present.shelf.spec.filters,
-      schema: state.present.dataset.schema,
+      spec: presentUndoableState.shelf.spec,
```
Comment (kanitw) [inline src/components/encoding-pane/index.tsx]: We should use a selector for dataset, shelf, shelfPreview for consistency and for reduce the amount of updates needed if we reshape the state in the future.

File: src/models/index.ts [hunk 5]
```diff
@@ -63,31 +75,52 @@ export interface SerializableState {
...
+  const persistentState = state.persistent;
+  const undoableState = state.undoable.present;
+
   const asSerializable = {
-    bookmark: state.bookmark,
-    config: state.config,
-    shelf: state.shelf,
-    shelfPreview: state.shelfPreview,
-    result: state.result,
+    bookmark: persistentState.bookmark,
```
Comment (kanitw) [inline src/models/index.ts]: just return it, no need for const

File: src/models/index.ts [hunk 6]
```diff
@@ -63,31 +75,52 @@ export interface SerializableState {
...
-export function fromSerializable(serializable: SerializableState): Readonly<StateBase> {
-  // We make a clone of this object to not modify the input param.
-  const deserialized = duplicate(serializable) as any;
+export function fromSerializable(serializable: SerializableState): Readonly<State> {
+  const {bookmark, config, shelf, shelfPreview, result, dataset, tableschema} = serializable;
 
-  // Add a schema object with a hydrated version of the table schema
-  deserialized.dataset.schema = new Schema(deserialized.tableschema);
-  delete deserialized.dataset.tableschema;
+  const deserialized: State = {
```
Comment (kanitw) [inline src/models/index.ts]: just return it, no need for const

File: src/selectors/shelf.test.ts [hunk 7]
```diff
@@ -8,45 +8,36 @@ describe('selectors/shelf', () => {
...
-            spec: {
-              ...DEFAULT_STATE.shelf.spec,
-              filters
-            }
-          }
-        }
-      };
+
+      const state = {... DEFAULT_STATE};
+      state.undoable.present.shelf.spec.filters = filters;
```
Comment (kanitw) [inline src/selectors/shelf.test.ts]: This is causing side effect to DEFAULT_STATE and may make other tests fail!

File: src/components/data-pane/data-selector.tsx (hunk 2 above)
Comment (felixcodes) [inline src/components/data-pane/data-selector.tsx]: update: we fixed this in one of the commits in this PR (to be pushed)

File: src/components/header/bookmark.tsx [hunk 8]
```diff
@@ -134,7 +134,7 @@ const BookmarkRenderer = CSSModules(BookmarkBase, styles);
 export const BookmarkPane = connect(
   (state: State) => {
     return {
-      bookmark: state.present.bookmark
+      bookmark: state.persistent.bookmark
```
Comment (kanitw) [inline src/components/header/bookmark.tsx]: selectBookmark?

File: src/selectors/result.ts [hunk 9]
```diff
@@ -1,31 +1,30 @@
...
 export const selectResult: {
-  [k in ResultType]?: Selector<StateWithHistory<StateBase>, Result>
+  [k in ResultType]?: Selector<State, Result>
 } = RESULT_TYPES.reduce((selectors, resultType) => {
-  selectors[resultType] = (state: State) => state.present.result[resultType];
+  selectors[resultType] = (state: State) => state.undoable.present.result[resultType];
   return selectors;
 }, {});
 
+// const selectMainResult = (state: State) => state.undoable.present.result.main;
```
Comment (kanitw) [inline src/selectors/result.ts]: Why adding this unused line?
